### PR TITLE
Update operator hierarchy

### PR DESCRIPTION
**Pull request by delyank**
This is addressing issue #4  open by Chak @chakshinglee . It addresses the operator update. This includes updating the operator hierarchy and the relaxation procedures. It also can resmooth the tentative prolongator respecting the new matrix.

### CODE REVIEW

**Comment by barker29** [comment]
Addresses Issue #4 .

File: amg/CMakeLists.txt [hunk 1]
```diff
@@ -107,15 +107,14 @@ list(APPEND TPL_LIBRARIES ${METIS_LIBRARY_PATH}/${METIS_LIB_NAME})
...
+          HINTS ${HYPRE_DIR}/include $ENV{HYPRE_DIR}/include)
 include_directories(${HYPRE_INCLUDE_PATH})
 set(HYPRE_LIB_NAME libHYPRE.a)
-find_library(HYPRE_LIB HYPRE
-  ${HYPRE_DIR}/lib
-  $ENV{HYPRE_DIR}/lib)
-list(APPEND TPL_LIBRARIES ${HYPRE_LIB})
+find_path(HYPRE_LIBRARY_PATH ${HYPRE_LIB_NAME}
+          HINTS  ${HYPRE_DIR}/lib $ENV{HYPRE_DIR}/lib)
+add_library(HYPRE_LIB STATIC IMPORTED)
```
Comment (barker29) [inline amg/CMakeLists.txt:115]: Why these changes? I don't have any particular problem with them (build still works fine for me). In particular is Hypre *always* built with static libs?

File: amg/inc/tg.hpp [hunk 2]
```diff
@@ -792,10 +754,10 @@ void tg_print_data(mfem::HypreParMatrix& A, const tg_data_t *tg_data)
               A.NNZ());
     SA_ASSERT(tg_data);
     SA_ASSERT(tg_data->interp);
-    SA_ASSERT( A.GetGlobalNumRows() ==  A.GetGlobalNumCols());
+    SA_ASSERT(A.GetGlobalNumRows() ==  A.GetGlobalNumCols());
     SA_ASSERT(tg_data->interp->GetGlobalNumRows() == A.GetGlobalNumRows());
     PROC_STR_STREAM << "Level 1 dimension: "
-                    << tg_data->interp->GetGlobalNumCols();
+                    << (tg_data->Ac?tg_data->Ac->GetGlobalNumCols():tg_data->interp->GetGlobalNumCols());
```
Comment (barker29) [inline amg/inc/tg.hpp]: Can we add some spaces and maybe break onto two lines? This is impossible to read.

File: amg/inc/adapt.hpp [hunk 3]
```diff
@@ -127,6 +128,31 @@ int adapt_approx_xbad(
     double *err, double *err0, int *executed_iters, double rtol/*=10e-12*/, double atol/*=10e-24*/,
     bool normalize/*=1*/, bool output=true);
 
+/*! \brief Updates operators and relaxation.
+
+    It updates the weighted l1-smoother in accordance to the new matrix \a A.
```
Comment (barker29) [inline amg/inc/adapt.hpp]: I feel like the documentation of this function and its multilevel sister could be more clear. These descriptions suggest the main thing that is changing is the *smoother*, where really the whole hierarchy changes. I recommend something like "Recomputes hierarchy while assuming spectral information is the same."

File: amg/test/mltest/mltest.cpp [hunk 4]
```diff
@@ -414,6 +414,10 @@ int main(int argc, char *argv[])
     args.AddOption(&identity_partition, "-ip", "--identity-partition",
                    "-nip", "--no-identity-partition",
                    "No agglomeration, one element per agglomerate.");
+    bool adapt = false;
+    args.AddOption(&adapt, "-ad", "--adapt",
```
Comment (barker29) [inline amg/test/mltest/mltest.cpp:418]: How hard would it be to add a test for this option? (I understand our test suite in this project is *very* incomplete, but if it is easy to add a quick test now is the time to do it.)

**Review by barker29 — APPROVED** [review]
This looks good! My few comments can be taken as suggestions, not demands.

File: amg/CMakeLists.txt (hunk 1 above)
Comment (delyank) [reply amg/CMakeLists.txt:115]: The only reason I can recall is to make it consistent with the way the other libraries are presented in the cmake files.

File: amg/inc/tg.hpp (hunk 2 above)
Comment (delyank) [reply amg/inc/tg.hpp]: Done

File: amg/inc/adapt.hpp (hunk 3 above)
Comment (delyank) [reply amg/inc/adapt.hpp]: Done

File: amg/test/mltest/mltest.cpp (hunk 4 above)
Comment (delyank) [reply amg/test/mltest/mltest.cpp:418]: Done
